cmd/kubelet/app/server_bootstrap_test.go

```diff
@@ -73,12 +73,14 @@ func Test_buildClientCertificateManager(t *testing.T) {
 	defer s.Close()
 
 	config1 := &restclient.Config{
-		UserAgent: "FirstClient",
-		Host:      s.URL,
+		UserAgent:     "FirstClient",
+		Host:          s.URL,
+		ContentConfig: restclient.ContentConfig{ContentType: runtime.ContentTypeJSON},
```

Comment:
I assume every place in this PR that is explicitly setting client config to JSON represents a use that is broken by the defaulting? Doesn't that indicate that existing uses of client-go will also be similarly broken?

Exposure of RESTClient, DoRaw, and ability to construct raw requests passing in []byte Body content makes it ~impossible to keep existing clients working while swapping out the default accept / content type :-/

Reply:
Based on my reading of these, all of them involved using a fake server that was making assertions about the request via JSON payloads.  Importantly, content type is only used when you pass in a `runtime.Object`.  `RESTClient` + `runtime.Object` to an API that doesn't support proto would break though, for example if you tried to use the `RESTClient` to interact with a custom resource (which I guess is technically possible?).  What if put this behind a client-go feature gate?

Reply:
> while swapping out the default accept / content type

Note that I didn't change the global default, just the default of the clients used for core APIs.

Reply:
yeah.... I'm trying to reason about what this does for the thousands of places that grab a RESTClient from the typed core APIs - https://github.com/search?q=%22.RESTClient%28%29%22+language%3AGo+-path%3Atest+-path%3Afake+-path%3Avendor&type=code&p=2

Previously, was any content type sent by default, or was it empty?

Reply:
> for example if you tried to use the `RESTClient` to interact with a custom resource (which I guess is technically possible?

Which people definitely do, see search results above.

I have a really hard time knowing how to make this safe if it modifies the behavior of the RESTClient() people can grab out of the typed clients

Reply:
(yet another reason why exposing RESTClient was a mistake - https://github.com/kubernetes/kubernetes/issues/26130)

Reply:
> Previously, was any content type sent by default, or was it empty?

I'll have to check where the defaulting happens, but IIRC it uses JSON automatically.

Reply:
> I have a really hard time knowing how to make this safe if it modifies the behavior of the RESTClient() people can grab out of the typed clients

If we think that the REST client is the only place that will get broken, I can try to add special casing around it to external keep callers from getting broken.

Reply:
@liggitt PTAL at the latest approach and associated tests that allow us to default to proto while keeping the REST client abuse working as-is.

Reply:
but users will still need to set the JSON contentype, right?

If they upgrade to the new clients the applications will fail until they change their code to set it to JSON

Reply:
You will only get a failure if all of the following conditions are met:

1. Upgrade to new client-go
2. Existing unit test
3. Unit test uses regular core client with no special content type config
4. Unit test uses core API like pod
5. Unit test uses http test server
6. Unit test http test server makes assertions on raw request payload (i.e. asserts pod contents as JSON)

If at any point the code tries to use a rest client directly, that will always end up using JSON unless the caller themselves had set protobuf.

Reply:
I was referring if someone in the wild is using the RESTClient() and have some frontend that expect json, and now receives proto, will it not break?

Reply:
That will not break because there is no way for that old code to be calling the new methods that opt-in to the protobuf.  The opt-in happens per request by calling specific methods on the request.  These methods did not exist in the old code.  You can see the test in the last commit that shows this case works without breaking callers.

Reply:
please bear with me :)

If I have this in my code 
`c.client.CoreV1().RESTClient().Get().AbsPath("/readyz").Do(context.TODO()).StatusCode(&code)` 

 when I vendor the new client-go v0.32.0 ... will that command not use proto instead of json as media type when I recompile?

Reply:
`c.client.CoreV1().RESTClient().Get().AbsPath("/readyz").Do(context.TODO()).StatusCode(&code)` will keep using JSON.

An explicit `UseProtobufAsDefault` call is needed to use proto:

`c.client.CoreV1().RESTClient().Get().UseProtobufAsDefault().AbsPath("/readyz").Do(context.TODO()).StatusCode(&code)`

Reply:
it seems a bit of a stretch adding public methods like that :/

Reply:
It is the cost we have to pay for exposing the rest client.  As ugly as those methods are, they aren't going to bother anyone by existing, and will allow the ecosystem to adopt proto by default for core APIs.  Personally, I think that is a fine trade off.

Reply:
It was a public rant, nothing else 🤷